File: crates/core/src/domain/palette.rs
/// The swatch names a project or group colour may take.
///
/// Stored values are these bare names, never literal colours — `src/lib/palette.ts`
/// resolves each to a `var(--color-swatch-*)` token at render, so a theme swap
/// recolours every project and group coherently. That file is a hand-maintained
/// mirror of this list; change one, change the other.
pub const SWATCHES: [&str; 8] = [
    "cyan", "gold", "amber", "rust", "violet", "green", "blue", "pink",
];

pub fn is_known_swatch(name: &str) -> bool {
    SWATCHES.contains(&name)
}

/// Exactly `#` followed by six hex digits — nothing else.
///
/// The strictness is the point, not pedantry: a stored colour is interpolated
/// into a `style` attribute by the frontend, and `style-src` still carries
/// `unsafe-inline`. Accepting only this shape is what guarantees a stored
/// value cannot carry arbitrary CSS. Shorthand (`#abc`), named CSS colours and
/// `rgb()` are all deliberately refused rather than normalised.
pub fn is_hex_literal(name: &str) -> bool {
    let Some(digits) = name.strip_prefix('#') else {
        return false;
    };
    digits.len() == 6 && digits.bytes().all(|b| b.is_ascii_hexdigit())
}

/// What a project or a group may store: a palette name, or a hex literal from
/// the colour picker.
///
/// A palette name is still the better answer — it is what lets a future theme
/// recolour everything coherently — but that is a reason to offer the palette
/// first, not a reason to refuse a colour someone actually wants.
pub fn is_valid_color(name: &str) -> bool {
    is_known_swatch(name) || is_hex_literal(name)
}

File: src/lib/palette.ts
// Hand-maintained mirror of crates/core/src/domain/palette.rs. That file is
// the source of truth for which names are storable; change one, change the
// other. Nothing checks this automatically.
//
// A stored colour is a bare name ("gold"), never a literal, so a future theme
// plugin can recolour every project and group coherently by overriding the
// @theme tokens in app.css. Resolving an unknown name to neutral rather than
// throwing is the same ignore-what-you-do-not-know rule the --json contract
// uses: a database written by a newer build must still render.
export const SWATCHES = [
  "cyan",
  "gold",
  "amber",
  "rust",
  "violet",
  "green",
  "blue",
  "pink",
] as const;

export type Swatch = (typeof SWATCHES)[number];

const NEUTRAL = "var(--color-phos-faint)";

export function isSwatch(name: unknown): name is Swatch {
  return typeof name === "string" && (SWATCHES as readonly string[]).includes(name);
}

// Returns a CSS value, not a colour literal. The return is always one of a
// fixed set of `var(--color-*)` strings, which is what makes it safe to
// interpolate into a `style` attribute — a stored colour name can never
// smuggle arbitrary CSS through it.
export function swatchVar(name: string | null | undefined): string {
  return isSwatch(name) ? `var(--color-swatch-${name})` : NEUTRAL;
}

// The colour a project's mark takes: its own, else its group's, else neutral.
//
// Inheriting the group's colour is what makes a project in a coloured group
// read as belonging to it at a glance, rather than sitting grey next to a
// coloured sidebar entry. The *pip* deliberately does not inherit — it is the
// secondary level, and if it did, "this project has its own colour" would be
// indistinguishable from "it inherits its group's".
export function markVar(
  projectColor: string | null | undefined,
  groupColor: string | null | undefined,
): string {
  return swatchVar(projectColor ?? groupColor);
}
